sql/preprocess/preprocess.go

```diff
@@ -0,0 +1,412 @@
+	text      strings.Builder
+	nesting   int
+	line, col int
+	err       error
+}
+
+func (p *processor) newTableName() string {
+	p.autoAliased++
+	return fmt.Sprintf("t%d", p.autoAliased)
+}
+
+func (p *processor) Process() error {
+	return p.processStatement()
+}
+
+func (p *processor) processStatement() error {
+	nesting := p.nesting
+
+	for {
+		tok := p.scan()
+
+		switch tok.Type {
+		case fromTok:
+			p.processed.WriteString(tok.Text)
+			if err := p.processTableRefs(); err != nil {
+				return err
+			}
+		case leftParenTok:
+			p.processed.WriteString(tok.Text)
+			p.nesting++
+			if err := p.processStatement(); err != nil {
+				return err
+			}
+		case rightParenTok:
+			p.processed.WriteString(tok.Text)
+			p.nesting--
+			if p.nesting == nesting {
+				return nil
+			}
+		case eofTok:
+			return nil
+		default:
+			p.processed.WriteString(tok.Text)
+		}
+	}
+}
+
+func (p *processor) processTableRefs() error {
+	hasReadName := false
+	nesting := p.nesting
+
+	for {
+		tok := p.scan()
+
+		switch tok.Type {
+		// check for table name references
+		case textTok:
+			if err := p.processTableRef(tok.Text); err != nil {
+				return err
+			}
+
+		default:
+			// default to writing any non-empty token text
+			if tok.Text != "" {
+				p.processed.WriteString(tok.Text)
+			}
+
+			switch tok.Type {
+			case commaTok:
+				if !hasReadName {
+					return fmt.Errorf("encountered '%s' before table name", tok.Text)
+				}
+				hasReadName = false
+
+			case leftParenTok:
+				p.nesting++
+				if err := p.processStatement(); err != nil {
+					return err
+				}
+			case rightParenTok:
+				p.nesting--
+				if p.nesting == nesting {
+					return nil
+				}
+				// pass an empty string to skip mapping, but still check for table alias
+				p.processTableRef("")
+
+			case asTok:
+			case whereTok, groupTok, havingTok, limitTok, orderTok, unionTok:
+				return nil
+			case eofTok:
+				return nil
+			default:
+			}
+		}
+	}
+}
+
+// processTableRef adds a reference to the map, writes a legal name, and
+// adds a table alias if one does not exist
+func (p *processor) processTableRef(text string) error {
+	if text != "" {
+		name := toLegalName(text)
+		p.mapping[name] = text
+		p.processed.WriteString(name)
+	}
+
+	alias := ""
+
+	for {
+		tok := p.scan()
+
+		switch tok.Type {
+		case whitespaceTok:
+			p.processed.WriteString(tok.Text)
+		case asTok:
+			p.processed.WriteString(tok.Text)
+		case textTok:
+			alias = tok.Text
+			if _, ok := p.aliases[alias]; ok {
+				return fmt.Errorf("duplicate reference alias '%s'", alias)
+			}
+			p.processed.WriteString(tok.Text)
+		default:
+			if alias == "" {
+				return fmt.Errorf("alias is required for reference '%s'", text)
+			}
+
+			switch tok.Type {
+			case commaTok, whitespaceTok:
+				p.processed.WriteString(tok.Text)
+			default:
+				p.unscan(tok)
+			}
+			return nil
+		}
+	}
+}
+
+func toLegalName(refStr string) string {
+	refStr = strings.Replace(refStr, "@", "_at_", 1)
```

Comment:
Think we still don't handle spaces and numeric names well here.

Reply:
On the other hand we can have the user ensure a well formated query instead of forcing a convention on to them or automagically swapping out too many things.